libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts

```diff
@@ -1,37 +1,49 @@
-import { BadRequestException, Injectable } from '@nestjs/common';
+import { Injectable } from '@nestjs/common';
 import { PreferencesEntity, PreferencesRepository } from '@novu/dal';
 import {
-  buildWorkflowPreferences,
   PreferencesTypeEnum,
+  WorkflowPreferences,
   WorkflowPreferencesPartial,
 } from '@novu/shared';
-import { UpsertPreferencesCommand } from './upsert-preferences.command';
 import { UpsertWorkflowPreferencesCommand } from './upsert-workflow-preferences.command';
 import { UpsertSubscriberGlobalPreferencesCommand } from './upsert-subscriber-global-preferences.command';
 import { UpsertSubscriberWorkflowPreferencesCommand } from './upsert-subscriber-workflow-preferences.command';
 import { UpsertUserWorkflowPreferencesCommand } from './upsert-user-workflow-preferences.command';
 import { deepMerge } from '../../utils';
 
+export type WorkflowPreferencesFull = Omit<PreferencesEntity, 'preferences'> & {
+  preferences: WorkflowPreferences;
+};
+
+type UpsertPreferencesCommand = Omit<
+  Partial<
+    UpsertWorkflowPreferencesCommand &
+      UpsertSubscriberGlobalPreferencesCommand &
+      UpsertSubscriberWorkflowPreferencesCommand &
+      UpsertUserWorkflowPreferencesCommand
+  >,
+  'preferences'
+> & {
+  organizationId: string;
+  environmentId: string;
+  type: PreferencesTypeEnum;
+  preferences: WorkflowPreferencesPartial;
+};
+
 @Injectable()
 export class UpsertPreferences {
   constructor(private preferencesRepository: PreferencesRepository) {}
 
   public async upsertWorkflowPreferences(
     command: UpsertWorkflowPreferencesCommand,
-  ) {
-    /*
-     * Only Workflow Preferences need to be built with default values to ensure
-     * there is always a value to fall back to during preference merging.
-     */
-    const builtPreferences = buildWorkflowPreferences(command.preferences);
-
+  ): Promise<WorkflowPreferencesFull> {
     return this.upsert({
       templateId: command.templateId,
       environmentId: command.environmentId,
       organizationId: command.organizationId,
-      preferences: builtPreferences,
+      preferences: command.preferences,
       type: PreferencesTypeEnum.WORKFLOW_RESOURCE,
-    });
+    }) as Promise<WorkflowPreferencesFull>;
```

Comment:
Not the most elegant, but it does the job for now without introducing a lot more generics for response types.